### ISSUE TITLE
Add support for multiple configurations in package.json

### ISSUE THREAD

**Issue by mathisonian** [idyll-lang/idyll#640]
**Is your feature request related to a problem? Please describe.**
It is relatively common to want to have two different idyll configurations, one for local development, and one for your production server. 

**Describe the solution you'd like**
There should be a way to specify multiple configurations in the `package.json` file. Something like this could work:

```js
{
  "idyll": [
     ["myDevConfig", { ...configuration options...}],
     ["myProductionConfig", { ...configuration options...}]
  ]
}
```

There would also need to be a way to specify which config you want to use. Something like `idyll --config myDevConfig` or `idyll --env myProductionConfig`. It should default to the first entry if none is specified, and clearly log which config is being used.

**Comment by mathisonian** [comment]
This should be added once https://github.com/idyll-lang/idyll/pull/606 is merged because I believe that will make the implementation a bit easier.

**Comment by rogerfitz** [comment]
Hey I'm starting to work on this. How should it handle multiple configuration files? Say the local config as #606 prioritizes doesn't list dev dependencies but the top level package.json does and dev config was specified via commandline. Is it OK to fail that or should it look for the first valid match when going through priorities? I would think just fail it with an error message like no dev dependencies found in confix file /x/y/z? Could also add an argument to let the user override default error/no error behavior here

**Comment by mathisonian** [comment]
Awesome @rogerfitz! I think failing that with an error message makes sense, its probably better to make things explicit here rather than pass it through and lead to a potentially unexpected configuration. So say we have the following situations:

```
-- enclosing-folder
----  package.json (has dev/prod config)
----  idyll-project
------ package.json (has dev/prod config)
```

^ in this case no error should be thrown, the `package.json` inside of `idyll-project` should take precedence but anything set in the outer `package.json` should be used if it wasn't set by the inner one

```
-- enclosing-folder
----  package.json (has dev/prod config)
----  idyll-project
------ package.json (doesn't have separate dev/prod config)
```

^ in this case an error should be thrown

```
-- enclosing-folder
----  package.json  (doesn't have separate dev/prod config)
----  idyll-project
------ package.json (has dev/prod config)
```

^ in this case no error should be thrown, the `package.json` inside of `idyll-project` should take precedence but property set in the outer package.json is used if isn't set in the inner one



I think this seems like a reasonable pattern to follow, but open to other suggestions. If in doubt prefer making things explicit and throw an error asking the user to fix something.

**Comment by rogerfitz** [comment]
Sounds good! Will go with that

**Comment by rogerfitz** [comment]
Hey sorry for the delay on this. How do I go about running my modified version of idyll cli code? I was following along here https://github.com/idyll-lang/idyll/blob/master/CONTRIBUTING.md#testing-changes but ran into an issue with build not working. Is cli a special package that doesn't need build? Or what's the best way to test?

I can run `cd packages/idyll-cli && yarn link`  but when I go to run idyll to test the idyll-cli changes it doesn't look like they are being picked up.

If it helps my idyll is pointing to /usr/bin/idyll, I did do a `npm install -g idyll`
```
ubuntu@DESKTOP-ILR8ANI:~/matteo_dev/t/my-idyll-post$ which idyll
/usr/bin/idyll
```

**Comment by mathisonian** [comment]
@rogerfitz you can do the following to test the local copy when building an idyll post:
- in `packages/idyll-cli && yarn link` (you've already done this it. seems)
- in `my-idyll-post` run `yarn link idyll` to use the linked version of idyll in that directory
- in `my-idyll-post` add `"scripts": { "start": "idyll" }` to `packages.json`
- in `my-idyll-post` run `yarn run start`, this will run the local copy of the idyll command line tool

Note that won't work if you just run `idyll` because that will still be using the globally installed version from `npm`. Only by running it via an npm script can you force yarn/npm to use the linked version of a command-line command

**Comment by rogerfitz** [comment]
Working thanks!

**Comment by rogerfitz** [comment]
Any thoughts on how best to create tests for this one? I started by mirroring idyll-cli/basic-projct but I'm not sure how to put everything in one test.js and run multiple command line envs in it. Like test command line arg ``--env prod``.  
   
I've got a base case here and was thinking of changing the idyll object to be a dictionary. https://github.com/rogerfitz/idyll/blob/master/packages/idyll-cli/test/env-options/test.js#L51. To do the nested test I was going to build a multi post idyll project and copy that over to src. Is that the best way to go about everything or am I making this harder than it needs to be?

Also, does anyone work on idyll in windows? I ran into some issues with yarn and husky but was able to get past them and could document it

**Comment by mathisonian** [comment]
Hey @rogerfitz- 

re: windows, yes it is expected that it should work. We run all of the tests on windows via the CI and ensure that they pass, although most users are not on windows so if there are issue it would be great to report them so we can fix / provide workarounds.

For testing, does it read the environment just from the command line args, or can you pass it in directly to the `Idyll(options)` command as well? If it supports the latter, I'd expect you could keep the same general structure and just have tests like:

```
test('no env provided', () => {
   const idyll = Idyll(...);

   expect(idyll.getOptions()).toEqual(...)
})

test('my-my provided', () => {
   const idyll = Idyll({..., env: 'my-env' });

   expect(idyll.getOptions()).toEqual(...)
})
```

since you are just testing that the resulting options are correct I don't think its necessary to test the build step as well for this feature, so can likely keep it relatively simple if that makes sense to you.

**Comment by rogerfitz** [comment]
Thanks that helped a lot. I created a PR and could discuss more of final details there or here. 

After this I can take a look at those windows issues more and see if it's more of a me problem or something could use tweaking in install steps. This was the main thing I had to do https://github.com/typicode/husky/issues/749#issuecomment-691531840 On second read it looks like I might be able to fix things by reinstalling differently. I tried running windows first but switched to WSL2 running Ubuntu on Windows 10 and had better luck

**Comment by mathisonian** [comment]
I've found the husky hooks to be fairly annoying as well, and occasionally flaky so i end up doing `git push origin <my-branch> --no-verify` to skip. I've been thinking about removing completely - we can discuss on a separate issue

**Comment by mathisonian** [comment]
Closed in #701 thanks again @rogerfitz!
